### PR TITLE
Better dead letter queues

### PR DESCRIPTION
**Pull request by BjRo**
Open TODOs

- [x] Check behaviour of dead letter cycle. Are messages really added at the end? **Yes, they are. Verified by publishing from rabbit web interface into both queues. Messages are added at the end of the original queue from the DL queue**
- [x] Check what happens when both queue and policy specify a message ttl? Does the smallest one win? **The smallest ttl wins if both policy and queue define a message-ttl. Verified on local rabbit instance**
- [x] Update documentation

### CODE REVIEW

File: lib/beetle/dead_letter_queue.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+      def set_dead_letter_queues!(servers, queue_name, options={})
+        servers.each { |server| set_dead_letter_queue!(server, queue_name, options) }
+      end
+
+      def set_dead_letter_queue!(server, queue_name, options={})
+        raise ArgumentError.new("server missing")     if server.blank?
+        raise ArgumentError.new("queue name missing") if queue_name.blank?
+
+        request_url = URI("http://#{server}/api/policies/%2F/#{queue_name}_policy")
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: We should not hardcode the vhost (/ aka %2F) here, but use the one from the config.

File: lib/beetle/dead_letter_queue.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
+require 'net/http'
+require 'json'
+
+module Beetle
+  class DeadLetterQueue
+    READ_TIMEOUT = 3 #seconds
+    DEFAULT_DEAD_LETTER_MSG_TTL = 1000 #1 second
+    RABBIT_API_PORT = 15672
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: Let's move them to the config.

File: lib/beetle/dead_letter_queue.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    class FailedRabbitRequest < StandardError; end
+
+    class << self
+      def bind_dead_letter_queues!(channel, servers, target_queue, creation_keys = {})
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: As discussed: Let's review the incoming creation_keys, if we have to handle/ignore e.g. per message ttl.

File: lib/beetle/dead_letter_queue.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        set_dead_letter_queues!(servers, dead_letter_queue_name,
+                                :message_ttl => DeadLetterQueue::DEFAULT_DEAD_LETTER_MSG_TTL,
+                                :routing_key => target_queue)
+      end
+
+      def set_dead_letter_queues!(servers, queue_name, options={})
+        servers.each { |server| set_dead_letter_queue!(server, queue_name, options) }
+      end
+
+      def set_dead_letter_queue!(server, queue_name, options={})
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: Better name: `def set_dead_letter_policy!`

File: lib/beetle/dead_letter_queue.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        logger.debug("Beetle: setting #{dead_letter_queue_name} as dead letter queue of #{target_queue} on all servers")
+        set_dead_letter_queues!(servers, target_queue)
+
+        logger.debug("Beetle: setting #{target_queue} as dead letter queue of #{dead_letter_queue_name} on all servers")
+        set_dead_letter_queues!(servers, dead_letter_queue_name,
+                                :message_ttl => DeadLetterQueue::DEFAULT_DEAD_LETTER_MSG_TTL,
+                                :routing_key => target_queue)
+      end
+
+      def set_dead_letter_queues!(servers, queue_name, options={})
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: Better name: `def set_dead_letter_policies!`

File: lib/beetle/dead_letter_queue.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      def dead_letter_routing_key(queue_name, options)
+        options.fetch(:routing_key) { dead_letter_queue_name(queue_name) }
+      end
+
+      def dead_letter_queue_name(queue_name)
+        "#{queue_name}_dead_letter"
+      end
+
+      def run_rabbit_http_request(uri, request, &block)
+        request.basic_auth("guest", "guest")
```
Comment (boosty) [inline lib/beetle/dead_letter_queue.rb]: Move to config.

File: beetle.gemspec [hunk 7]
```diff
@@ -32,4 +32,6 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency("activesupport",           [">= 2.3.4"])
   s.add_runtime_dependency("eventmachine_httpserver", [">= 0.2.1"])
   s.add_runtime_dependency("daemons",                 [">= 1.0.10"])
+
+  s.add_development_dependency("webmock", ["1.21.0"])
```
Comment (boosty) [inline beetle.gemspec]: Needs exact version?

File: examples/attempts.rb [hunk 8]
```diff
@@ -35,6 +38,12 @@ class Handler < Beetle::Handler
 
   # called when the handler receives the message - fail everytime
   def process
+    logger.info "received message with routing key: #{message.routing_key}"
+    death = message.header.attributes[:headers]["x-death"]
+    if death
+      logger.info "X-DEATH: Message is comming back from dead letter queue (#{death.first["count"]})"
```
Comment (boosty) [inline examples/attempts.rb]: comming -> coming

File: lib/beetle/dead_lettering.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      logger.debug("Beetle: creating dead letter queue #{dead_letter_queue_name} with opts: #{creation_keys.inspect}")
+      dead_letter_queue = channel.queue(dead_letter_queue_name, creation_keys)
+
+      logger.debug("Beetle: setting #{dead_letter_queue_name} as dead letter queue of #{target_queue} on all servers")
+      set_dead_letter_policies!(servers, target_queue)
+
+      logger.debug("Beetle: setting #{target_queue} as dead letter queue of #{dead_letter_queue_name} on all servers")
+      set_dead_letter_policies!(servers, dead_letter_queue_name,
+                              :message_ttl => @config.dead_lettering_msg_ttl,
+                              :routing_key => target_queue)
```
Comment (boosty) [inline lib/beetle/dead_lettering.rb]: Indentation is a bit off here.

**Comment by BjRo** [comment]
Merged via f46c213cf9ce2607d8bf456d8ccf674afcf44cf5

**Pull request closed without merging**